### ISSUE TITLE
performance observations while interacting with a 23k-tip tree

### ISSUE THREAD

**Issue by jameshadfield** [nextstrain/auspice#1878]
A summary of observations while rendering & interacting with a huge tree. Ideally we'd like to have Auspice be performant on such big trees, but also this is a simple way to flag up slow parts of the code. The timings are just to indicate bottlenecks - while they are reasonably consistent for me they'll be different for you. (This issue is mainly for me to keep track of things in a public-facing way. If you'd like to work on anything please spin it out into a new issue.) 

### Setup: 

Dataset: `groups/blab-private/ncov/north-america/usa/washington/king-county/large?c=ns_kc&d=tree&p=full` (i.e. only tree panel rendering). 23k tips. The dataset is only 45MB, so lots of nodes but (relatively) few mutations (45MB minimised but uncompressed). 

Auspice 2.59.1 at master branch (35e4ca69) running in production mode with the following changes:
```diff
webpack.config.js
-    devtool: !devMode ? undefined : "eval-source-map",
+    devtool: "eval-source-map",
-      minimize: !devMode,
+      minimize: false,
```

Chrome 130.0 running in incognito mode on an Apple M3 Pro chip. GC triggered before each snapshot was taken.

## Test 1:

Filter to `f_region=South%20America` (n=259 tips selected). Start a snapshot then inactivate the filter, immediately hover over a tip so we can see when that listenerr will fire/render. This shares a lot of similarities to dragging the date range / animating but is much easier to reproduce consistently & isolate. `Trace-20241026T131320.json`

**Heap Size:** 230MB to 330Mb.
**Timings:**
1. 666ms JS "Event: click". This is our code & d3 code.
2. 240ms JS "Animation frame fired' (triggered from (1)). This is mostly d3 code.
3. 115ms recalculate style (triggered from (2))
4. 40ms (pre-paint, paint, layerize)  (no visual changes - looks identical)
5. 1170ms Commit 
   * During this time, partially-presented frames are rendered at 60fps
   * About ~100ms into the Commit cycle we get the tips all shown but not all the branches are correct. So it looks like partial d3 updates have been rendered?
   * All the other partially-presented frames look identical
6. <1ms Event: mouseover 
7. 3ms Recalculate style, layout, pre-paint
   * NOTE: no paint here
8. 75ms JS "Animation frame fired" (triggered from (1))
9. 80ms recalculate style (triggered from (6)) 
10. 45ms (pre-paint, paint, layerize)
   * This is the first screenshot where the tree is "correct".
   * We also have the tip hover showing here in the same frame as the tree becomes correctly rendered.

TLDR: Total time to partial tree update painted: ~1250ms. Total time to tree fully painted & responsive: 2500ms.

#### Don't try to animate things

Improvements: huge; should be an easy win if we can make a heuristic about when to animate vs not. Changing ``genericSelectAndModify`` to no longer have`.transition().duration(transitionTime)` provides a huge improvement.

* In the timings above, (1) is reduced by 100ms. Most of this saving is a reduction of time spent in `genericSelectAndModify`. 
* We completely get rid of steps (2,8,9,10)
* The paint during (5: commit) is the correct one - i.e the tree is correctly rendered. Note the commit still takes ~1200ms with the paint coming very early on in this phase. My understanding of what the commit phase actually does is limited.

Note: Achieving the same thing by using a `transitionTime=0` and thus calling `timerFlush()` is worse. Step 1 takes ~900ms with `timerFlush` taking ~300ms. 

#### Recalculating entropy, despite the entropy panel not being shown

Improvements: large. `calcEntropyInView` takes 295ms. The entropy panel's not even rendered! 

We should add a concept of "thing X doesn't need to be recalculated" - e.g. when the entropy panel's off the screen or not even on. The data can then be recalculated when needed.

Even when the entropy panel is on the screen, it can have delayed updating. Seems perfect for web-workers - e.g. show a grey overlay on the panel itself to indicate data out-of-date, recalculate things in the worker, then trigger the redux update. (Aside: we don't have any web workers in Auspice, but there are many situations they would be useful. [This blog post](https://developers.redhat.com/blog/2014/05/20/communicating-large-objects-with-web-workers-in-javascript) seems a good resource on how workers may access the tree structure.)

#### Unnecessary recalculation of filtering (sidebar) options
Improvements: medium, but easy wins. Total time: 41ms. The sidebar filtering component `<FilterData>` updates on any filter change (not date range). Each time it calls its [makeOptions](https://github.com/nextstrain/auspice/tree/35e4ca6961f1e0cd9a36cf06d8713e459d0dfff1) method which recomputes all options (to show in the dropdown), even though most of them are unchanged. Calling `collectGenotypeStates` takes 30ms, but the result of this doesn't ever change so is completely unnecessary.

Note: the options are calculated every render time, but we're debouncing the callback which hands them to  `<AsyncSelect>`.  Can we just defer the calculation of `options` to this callback?

#### Expensive tooltips 
The unmounting of the react tooltip (shown when hovering on the filter badge) takes 40ms and then it immediately mounts again (40ms). These aren't deferred / put onto the event loop for some reason (I don't quite get this). Probably hard to improve but seems egregious.
